source/reduce/remove_block_reduction_opportunity.cpp

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "remove_block_reduction_opportunity.h"
+
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace reduce {
+
+using namespace opt;
+
+RemoveBlockReductionOpportunity::RemoveBlockReductionOpportunity(
+    Function* function, BasicBlock* block)
+    : function_(function), block_(block) {
+  // precondition:
+  assert(block_->begin() != block_->end() &&
+         block_->begin()->context()->get_def_use_mgr()->NumUsers(
+             block_->id()) == 0 &&
+         "RemoveBlockReductionOpportunity block must have 0 references");
+}
+
+bool RemoveBlockReductionOpportunity::PreconditionHolds() {
+  // Removing other blocks cannot disable this opportunity.
+  return true;
+}
+
+void RemoveBlockReductionOpportunity::Apply() {
+  // We need an iterator pointing to the block, hence the loop.
+  for (auto bi = function_->begin(); bi != function_->end(); ++bi) {
```

Comment:
If I have blocks A and B that are unreachable and A branches to B and B uses some values defined in A, you may leave dangling uses in B I think. Since B is referenced it wouldn't be removed. Consider replacing uses in B of values defined in A with undef.

Reply:
Good point.  Rather than actually replacing uses, I think it would be better to rely on existing reduction passes that remove uses to get rid of them (I'm in favour of keeping passes as separate as possible unless we find that it's really necessary to have one pass do work that another pass can also do.)

Reply:
So would you check that none of the instruction in A has any uses?

Reply:
Yes, although I suppose uses from inside the block are OK. I have attempted to do this. Still need to add a unit test.

Reply:
OK. Unit tests added.